### ISSUE TITLE
ComputationGraph ReshapePreprocessor Error! Multiple Input INDArray.

### ISSUE THREAD

**Issue by andyxzq** [deeplearning4j/deeplearning4j#5311]
java.lang.IllegalStateException: Input shape [1, 10, 1] and output shape[1] do not match

	at org.deeplearning4j.nn.modelimport.keras.preprocessors.ReshapePreprocessor.preProcess(ReshapePreprocessor.java:100)
	at org.deeplearning4j.nn.graph.vertex.impl.PreprocessorVertex.doForward(PreprocessorVertex.java:62)
	at org.deeplearning4j.nn.graph.ComputationGraph.outputOfLayersDetached(ComputationGraph.java:2156)
	at org.deeplearning4j.nn.graph.ComputationGraph.output(ComputationGraph.java:1638)
	at org.deeplearning4j.nn.graph.ComputationGraph.outputSingle(ComputationGraph.java:1621)
	at org.deeplearning4j.nn.graph.ComputationGraph.outputSingle(ComputationGraph.java:1607)
	at org.deeplearning4j.nn.graph.ComputationGraph.outputSingle(ComputationGraph.java:1551)
	at com.ximalaya.brain.deep.keras.KerasModelTest.testModel(KerasModelTest.java:125)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:160)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)

```
ComputationGraph cg = KerasModelImport.importKerasModelAndWeights(
                "/Users/nali/Documents/deepfm.h5"
        );
 System.out.println(cg.summary());
 INDArray[] input = new NDArray[10];
 for(int i = 0; i < 10; i++) {
     input[i] = Nd4j.ones(1, 1);
 }
 cg.outputSingle(input);
```

```
model summary:
==========================================================================================================================================================================================================================================================
VertexName (VertexType)                 nIn,nOut  TotalParams ParamsShape                             Vertex Inputs                 
==========================================================================================================================================================================================================================================================
input_0 (InputVertex)                   -,-       -           -                                       -                             
input_1 (InputVertex)                   -,-       -           -                                       -                             
input_2 (InputVertex)                   -,-       -           -                                       -                             
input_3 (InputVertex)                   -,-       -           -                                       -                             
input_4 (InputVertex)                   -,-       -           -                                       -                             
input_5 (InputVertex)                   -,-       -           -                                       -                             
input_6 (InputVertex)                   -,-       -           -                                       -                             
input_7 (InputVertex)                   -,-       -           -                                       -                             
input_8 (InputVertex)                   -,-       -           -                                       -                             
input_9 (InputVertex)                   -,-       -           -                                       -                             
embed_0 (EmbeddingSequenceLayer)        6,10      60          W:{6,10}                                [input_0]                     
linear_0 (EmbeddingSequenceLayer)       6,1       6           W:{6,1}                                 [input_0]                     
linear_1 (EmbeddingSequenceLayer)       41,1      41          W:{41,1}                                [input_1]                     
embed_1 (EmbeddingSequenceLayer)        41,10     410         W:{41,10}                               [input_1]                     
linear_2 (EmbeddingSequenceLayer)       8,1       8           W:{8,1}                                 [input_2]                     
embed_2 (EmbeddingSequenceLayer)        8,10      80          W:{8,10}                                [input_2]                     
linear_3 (EmbeddingSequenceLayer)       3,1       3           W:{3,1}                                 [input_3]                     
embed_3 (EmbeddingSequenceLayer)        3,10      30          W:{3,10}                                [input_3]                     
linear_4 (EmbeddingSequenceLayer)       24,1      24          W:{24,1}                                [input_4]                     
embed_4 (EmbeddingSequenceLayer)        24,10     240         W:{24,10}                               [input_4]                     
linear_5 (EmbeddingSequenceLayer)       11,1      11          W:{11,1}                                [input_5]                     
embed_5 (EmbeddingSequenceLayer)        11,10     110         W:{11,10}                               [input_5]                     
embed_6 (EmbeddingSequenceLayer)        6,10      60          W:{6,10}                                [input_6]                     
linear_6 (EmbeddingSequenceLayer)       6,1       6           W:{6,1}                                 [input_6]                     
embed_7 (EmbeddingSequenceLayer)        41,10     410         W:{41,10}                               [input_7]                     
linear_7 (EmbeddingSequenceLayer)       41,1      41          W:{41,1}                                [input_7]                     
embed_8 (EmbeddingSequenceLayer)        11,10     110         W:{11,10}                               [input_8]                     
linear_8 (EmbeddingSequenceLayer)       11,1      11          W:{11,1}                                [input_8]                     
embed_9 (EmbeddingSequenceLayer)        245,10    2450        W:{245,10}                              [input_9]                     
linear_9 (EmbeddingSequenceLayer)       245,1     245         W:{245,1}                               [input_9]                     
reshape_11 (PreprocessorVertex)         -,-       -           -                                       [embed_0]                     
reshape_1 (PreprocessorVertex)          -,-       -           -                                       [linear_0]                    
reshape_2 (PreprocessorVertex)          -,-       -           -                                       [linear_1]                    
reshape_12 (PreprocessorVertex)         -,-       -           -                                       [embed_1]                     
reshape_3 (PreprocessorVertex)          -,-       -           -                                       [linear_2]                    
reshape_13 (PreprocessorVertex)         -,-       -           -                                       [embed_2]                     
reshape_4 (PreprocessorVertex)          -,-       -           -                                       [linear_3]                    
reshape_14 (PreprocessorVertex)         -,-       -           -                                       [embed_3]                     
reshape_5 (PreprocessorVertex)          -,-       -           -                                       [linear_4]                    
reshape_15 (PreprocessorVertex)         -,-       -           -                                       [embed_4]                     
reshape_6 (PreprocessorVertex)          -,-       -           -                                       [linear_5]                    
reshape_16 (PreprocessorVertex)         -,-       -           -                                       [embed_5]                     
reshape_17 (PreprocessorVertex)         -,-       -           -                                       [embed_6]                     
reshape_7 (PreprocessorVertex)          -,-       -           -                                       [linear_6]                    
reshape_18 (PreprocessorVertex)         -,-       -           -                                       [embed_7]                     
reshape_8 (PreprocessorVertex)          -,-       -           -                                       [linear_7]                    
reshape_19 (PreprocessorVertex)         -,-       -           -                                       [embed_8]                     
reshape_9 (PreprocessorVertex)          -,-       -           -                                       [linear_8]                    
reshape_20 (PreprocessorVertex)         -,-       -           -                                       [embed_9]                     
reshape_10 (PreprocessorVertex)         -,-       -           -                                       [linear_9]                    
multiply_2 (ElementWiseVertex)          -,-       -           -                                       [reshape_11, reshape_11]      
multiply_3 (ElementWiseVertex)          -,-       -           -                                       [reshape_12, reshape_12]      
multiply_4 (ElementWiseVertex)          -,-       -           -                                       [reshape_13, reshape_13]      
multiply_5 (ElementWiseVertex)          -,-       -           -                                       [reshape_14, reshape_14]      
multiply_6 (ElementWiseVertex)          -,-       -           -                                       [reshape_15, reshape_15]      
multiply_7 (ElementWiseVertex)          -,-       -           -                                       [reshape_16, reshape_16]      
multiply_8 (ElementWiseVertex)          -,-       -           -                                       [reshape_17, reshape_17]      
multiply_9 (ElementWiseVertex)          -,-       -           -                                       [reshape_18, reshape_18]      
multiply_10 (ElementWiseVertex)         -,-       -           -                                       [reshape_19, reshape_19]      
add_2 (ElementWiseVertex)               -,-       -           -                                       [reshape_11, reshape_12, reshape_13, reshape_14, reshape_15, reshape_16, reshape_17, reshape_18, reshape_19, reshape_20]
concatenate_1 (MergeVertex)             -,-       -           -                                       [reshape_11, reshape_12, reshape_13, reshape_14, reshape_15, reshape_16, reshape_17, reshape_18, reshape_19, reshape_20]
multiply_11 (ElementWiseVertex)         -,-       -           -                                       [reshape_20, reshape_20]      
add_1 (ElementWiseVertex)               -,-       -           -                                       [reshape_1, reshape_2, reshape_3, reshape_4, reshape_5, reshape_6, reshape_7, reshape_8, reshape_9, reshape_10]
multiply_1 (ElementWiseVertex)          -,-       -           -                                       [add_2, add_2]                
dense_1 (DenseLayer)                    100,200   20200       W:{100,200}, b:{1,200}                  [concatenate_1]               
add_3 (ElementWiseVertex)               -,-       -           -                                       [multiply_2, multiply_3, multiply_4, multiply_5, multiply_6, multiply_7, multiply_8, multiply_9, multiply_10, multiply_11]
batch_normalization_1 (BatchNormalization)1,1       4           gamma:{1,1}, beta:{1,1}, mean:{1,1}, var:{1,1}[add_1]                       
activation_1 (ActivationLayer)          -,-       0           -                                       [dense_1]                     
subtract_1 (ElementWiseVertex)          -,-       -           -                                       [multiply_1, add_3]           
dense_2 (DenseLayer)                    200,200   40200       W:{200,200}, b:{1,200}                  [activation_1]                
activation_2 (ActivationLayer)          -,-       0           -                                       [dense_2]                     
concatenate_2 (MergeVertex)             -,-       -           -                                       [batch_normalization_1, subtract_1, activation_2]
dense_3 (DenseLayer)                    211,1     212         W:{211,1}, b:{1,1}                      [concatenate_2]               
dense_3_loss (LossLayer)                -,-       0           -                                       [dense_3]   
```
Does my inputs have some problems?

**Comment by maxpumperla** [comment]
@andyxzq thanks for reporting this, will get back to you soon.

**Comment by maxpumperla** [comment]
@andyxzq can you please give me your keras model, i.e. your full `h5` file with model and weights? Alternatively a clean script generating this model would do.

Just from the summary output I imagine that the issue comes from the output of an `EmbeddingSequenceLayer`. For instance, I see that `embed_0` has `nOut=10`, which leads to an output shape of `[mb, 10, 1]`. For minibatch size mb=1 this could be the input to a `Reshape` layer that seems to fail. Output shape [1] means that you're trying to reshape to a scalar... not sure in which scenario this makes sense.

We had a few fixes regarding Reshape lately, but I can only tell you more if you give me your model.

**Comment by maxpumperla** [comment]
@andyxzq could you please provide me with you model?

**Comment by andyxzq** [comment]
My model is DeepFM, which concatnates the output of FM and multi-layer network. After embedding, the output is 3-dim and I need to reshape the output to 2-dim so that it can horizontally concatenate with the 2-dim output of FM.

**Comment by andyxzq** [comment]
@maxpumperla  My model is DeepFM, which concatnates the output of FM and multi-layer network. After embedding, the output is 3-dim and I need to reshape the output to 2-dim so that it can horizontally concatenate with the 2-dim output of FM.

**Comment by maxpumperla** [comment]
@andyxzq I meant actual code.

**Comment by agibsonccc** [comment]
Closing — ND4J/libnd4j core infrastructure has been significantly rewritten (new memory management, op execution engine, backend architecture). If this issue is still relevant with the upcoming release, please open a new issue.
